Gems/Atom/Feature/Common/Assets/Shaders/PostProcessing/MotionBlur.azsl

```diff
@@ -0,0 +1,176 @@
+{
+    return saturate(0.5f + float2(depthScale, -depthScale) * (sampleDepth - centerDepth));
+}
+
+float2 SpreadCmp(float offsetLen, float2 spreadLen, float pixelToSampleUnitsScale)
+{
+    return saturate(pixelToSampleUnitsScale * spreadLen - offsetLen + 1.0f);
+}
+
+float SampleWeight(float centerDepth, float sampleDepth, float offsetLen, float centerSpreadLen, float sampleSpreadLen, float pixelToSampleUnitsScale, float depthScale)
+{
+    float2 depthCmp = DepthCmp(centerDepth, sampleDepth, depthScale);
+    float2 spreadCmp = SpreadCmp(offsetLen, float2(centerSpreadLen, sampleSpreadLen), pixelToSampleUnitsScale);
+    return dot(depthCmp, spreadCmp);
+}
+
+
+uint2 ClampCoords(int2 coord)
+{
+    uint2 clampedCoords;
+    clampedCoords.x = uint(clamp(coord.x, 0, PassSrg::m_constants.m_outputSize.x - 1));
+    clampedCoords.y = uint(clamp(coord.y, 0, PassSrg::m_constants.m_outputSize.y - 1));
+    return clampedCoords;
+}
+
+float2 ToPixelVelocity(float2 velocity)
+{
+    velocity *= float2(PassSrg::m_constants.m_outputSize);
+    velocity *= PassSrg::m_constants.m_strength;
+    return velocity;
+}
+
+// This shader calculates motion blur effect from tile max speeds. 
+// See https://advances.realtimerendering.com/s2014/index.html#_NEXT_GENERATION_POST
+// for a detailed overview of the technique. 
+[numthreads(TILE_DIM_X, TILE_DIM_Y, 1)]
+void MainCS(
+    uint3 dispatchThreadID : SV_DispatchThreadID, 
+    uint3 groupID : SV_GroupID, 
+    uint groupIndex : SV_GroupIndex)
+{
+    uint2 pixelCoord = dispatchThreadID.xy;
+    if (pixelCoord.x >= PassSrg::m_constants.m_outputSize.x || pixelCoord.y >= PassSrg::m_constants.m_outputSize.y)
+    {
+        return;
+    }
+
+    float3 color = PassSrg::m_inputColor[pixelCoord].rgb;
+    PassSrg::m_outputColor[pixelCoord].rgb = color;
+
+    float depth = PassSrg::m_linearDepth[pixelCoord].r;
+
+    float2 tileVelocity = PassSrg::m_inputTileVelocity[pixelCoord / uint2(TILE_SIZE, TILE_SIZE)];
+    tileVelocity = ToPixelVelocity(tileVelocity);
+    float tileVelocityLength = length(tileVelocity);
+
+    // Omit effect for velocities that are too small
+    if (tileVelocityLength < 1)
+    {
+        return;
+    }
+
+    // Introduce some dithering according to the slides
+    int2 ditherCoord = int2(pixelCoord.x % 2, pixelCoord.y % 2);
+    float samplingDither = (-0.25 + 2.0 * 0.25 * ditherCoord.x) * (-1.0 + 2.0 * ditherCoord.y);
+
+    int iterationCount;
+    switch(o_sampleNumber){
+        case SampleNumber::Sample6:
+        {
+            iterationCount = 3;
+            break;
+        }
+        case SampleNumber::Sample14:
+        {
+            iterationCount = 7;
+            break;
+        }
+        case SampleNumber::Sample24:
+        {
+            iterationCount = 12;
+            break;
+        }
+        case SampleNumber::Sample36:
+        {
+            iterationCount = 18;
+            break;
+        }
+    }
```

Comment:
My apologies. I didnt see that the iterationCount is not hardcoded and passed in. My original comment was meant to simply add the keyword which i see is not possible. My recommendation for now is to remove this new logic around o_sampleNumber and also remove the unroll keyword below. We can add this shader option later with perf numbers to justify the change.

Reply:
Got it.